### DISCUSSION TITLE
Comprehension of IResourceRequestHandler and interception of requests

### DISCUSSION THREAD

**Discussion by ordinaryorange** [cefsharp/CefSharp#3904] [General]
I'm currently trying to do some request accounting. By that I mean tracking what requests are sent from the browser and what are received.

I'm overriding the `ResourceRequestHandler.OnBeforeResourceLoad` and `ResourceRequestHandler.OnResourceResponse` methods to track the requests and responses which mostly works, except for the following odd behaviour.

For a particular page I'm logging requests & responses on, it issues requests for some fonts.
In my logs from CefSharp I see 2 entries from the `OnBeforeResourceLoad` (for the same Url) and only 1 log entry from the `OnResourceResponse` method.

I have been running all traffic through a proxy, and when I inspect those logs, the proxy confirms only 1 request/response was processed.

To investigate further, I hooked up the `ResourceRequestHandler.OnResourceLoadComplete` for logging as well.
For every `OnBeforeResourceLoad` I see a corresponding log entry from the `OnResourceLoadComplete` method.

I have a suspicion that the first `OnBeforeResourceLoad` request is being handled by another subsystem possibly involved with a cache?, and the second then going out the network as a result of a failure of the first, and the `OnResourceResponse` method is bypassed for non network responses ???. (but I'm speculating here)

Am I wrong to expect a log from `OnResourceResponse` for every `OnBeforeResourceLoad` (excluding lost requests of course) ?
If I'm correct about the cache thing, how do I account cache responses, if `OnResourceResponse` is not supposed to be called for cache events ?

**Comment by amaitland** [comment]
> I have a suspicion that the first OnBeforeResourceLoad request is being handled by another subsystem possibly involved with a cache?, and the second then going out the network as a result of a failure of the first, and the OnResourceResponse method is bypassed for non network responses ???. (but I'm speculating here)

Check the log file and check  DevTools to see if there are any clues as to what's going on.

- https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#log-file
- https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#javascript-debugging

**Reply by ordinaryorange** [reply]
Turned logs to verbose, but nothing of interest that I can see. 
The url's in question are in the logs, but only associated with `NetworkDelegate::NotifyBeforeURLRequest` or `Will sniff content for mime type:` log entries. 
There are no log entries at all with anything related to `NotifyBeforeResponse` or `NotifyResourceComplete` which is odd given the verbose setting.

I was applying the settings `disable-application-cache` and `disable-session-storage`, so to test my theory I disabled this,  let the cache create on a first run, but the second still resulted in all the same calls. So that's not it.

No errors or anything related in the console logs.

Not quite sure where to go with the javascript debugging, I don't own the page in question.

Do you think the `OnResourceResponse` should be called for every `OnBeforeResourceLoad` related url ?

I can use the `OnResourceLoadComplete` as a fallback, but the `OnResourceResponse` would be nicer.

Any more ideas, please post and I'll try.

**Reply by amaitland** [reply]
> Not quite sure where to go with the javascript debugging, I don't own the page in question.

You can look at the Network tab in DevTools. You can also load the website in Chrome and load DevTools to compare. If it looks different in Chrome then it's probably a bug in CEF's network interceptor, there a few little quirks. 

> Do you think the OnResourceResponse should be called for every OnBeforeResourceLoad related url ?

Short answer no. I don't remember the exact sequence off the top of my head. You can have a look at the CEF source if you're interested https://github.com/chromiumembedded/cef/blob/a7bbd8a62bfc91b0d53eeef8d07b64a5ed719a5f/libcef/browser/net_service/resource_request_handler_wrapper.cc#L741

OnResourceLoadComplete is meant to be called even when shutdown is called as per https://github.com/chromiumembedded/cef/blob/a7bbd8a62bfc91b0d53eeef8d07b64a5ed719a5f/include/cef_resource_request_handler.h#L175

You can disable CEFs network interceptor and use the DevTools protocol if it looks like a bug in CEF.

**Reply by ordinaryorange** [reply]
Had a cursory look at the Cef source, but I got lost early, so gave up as not critical.

Checked out the network requests in devtool and there is definitely only 1 request, all looks the same in Chrome.

Copied one of the requests as a fetch, and then pasted it into console. `OnBeforeResourceLoad`, then `OnResourceResponse` then `OnResourceLoadComplete` fired. No extra logs. Whilst that is not conclusive, I am leaning to a Cef bug as you suggest.

I'll make do with `OnResourceResponse` and if that causes issues I'll go down the DevTools protocol path.

Thanks for you help and ideas.

**Reply by amaitland** [reply]
> Copied one of the requests as a fetch, and then pasted it into console.

Is the original request a fetch? I did see a strange case reported regarding fetch requests a little while ago. https://magpcss.org/ceforum/viewtopic.php?f=6&t=18725&p=50444

**Reply by ordinaryorange** [reply]
Just checked. No it is coming from a CSS stylesheet.
I even just compared all the properties of the `IRequest`'s for the same url. Everything matches bar the `Identifier`.
This one is beyond me.
